src/Graphing/Debug.hs

```diff
@@ -0,0 +1,103 @@
+{-# LANGUAGE RecordWildCards #-}
+
+module Graphing.Debug (
+  GraphDebug (toDebug),
+  debugGraph,
+  renderGvFile,
+) where
+
+import Algebra.Graph.Acyclic.AdjacencyMap qualified as Acyclic
+import Algebra.Graph.AdjacencyMap qualified as Cyclic
+import Control.Effect.Lift (Has, Lift, sendIO)
+import Data.Bifunctor (Bifunctor (bimap))
+import Data.Set (Set)
+import Data.Set qualified as Set
+import Data.String (IsString)
+import Data.Text (Text)
+import Data.Text qualified as Text
+import Data.Text.Extra (showT)
+import Data.Text.IO qualified as TIO
+import DepTypes (DepEnvironment, Dependency (..), VerConstraint (..))
+import Graphing (Graphing (unGraphing), Node (..))
+
+gvFileName :: (IsString a) => a
+gvFileName = "./fossa.debug.gv"
+
+class Ord a => GraphDebug a where
+  toDebug :: a -> Text
+
+instance GraphDebug a => GraphDebug (Node a) where
+  toDebug Root = "ROOT"
+  toDebug (Node a) = toDebug a
+
+instance GraphDebug Dependency where
+  toDebug Dependency{..} = Text.unwords [dependencyName, version, envs]
+    where
+      version = maybe "*" versionTxt dependencyVersion
+      envs = "[" <> envsToText dependencyEnvironments <> "]"
+
+envsToText :: Set DepEnvironment -> Text
+envsToText = Text.intercalate ", " . map showT . Set.toList
+
+versionTxt :: VerConstraint -> Text
+versionTxt = \case
+  CEq txt -> txt
+  CURI txt -> txt
+  CCompatible txt -> txt
+  CAnd vc vc' -> versionTxt vc <> "AND" <> versionTxt vc'
```

Comment:
[nit] This might be overkill for now, but with nested combinations of `AND` and `OR` it could get confusing to see how things associate. Maybe it would be worth adding explicit parens around the operands?

Reply:
For now, I don't think this is going to be a problem.  I haven't actually seen these used in quite some time, and never in JS.